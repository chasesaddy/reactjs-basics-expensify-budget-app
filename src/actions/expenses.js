import uuid from 'uuid';
import db from '../firebase/firebase';

// 
//// Expenses
// 

// ADD_EXPENSE

const addExpense = ( expense ) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = ( expenseData = {} ) => {
  return ( dispatch ) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return db.ref( 'expenses' ).push( expense ).then( ( ref ) => { 
      dispatch( addExpense( { 
        id: ref.key,
        ...expense
      } ) );
    } );
  };
};

// REMOVE_EXPENSE

const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

const startRemoveExpense = ( { id } = {} ) => {
  return ( dispatch ) => {
    return db.ref( `expenses/${ id }` ).remove().then( ( ref ) => { 
      dispatch( removeExpense( { id } ) );
    } );
  };
};

// EDIT_EXPENSE

const editExpense = ( id, updates ) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
const setExpenses = ( expenses ) => ( { 
  type: 'SET_EXPENSES',
  expenses
} );

const startSetExpenses = () => {
  return ( dispatch ) => {
    return db.ref( 'expenses' ).once( 'value' ).then( ( snapshot ) => { 
      const allExpenses = [];
      snapshot.forEach( ( solo ) => { 
        allExpenses.push( {
          id: solo.key,
          ...solo.val()
        } );
      } );      
      dispatch( setExpenses( allExpenses ) );
    } );
  };
};

export { addExpense, startAddExpense, removeExpense, startRemoveExpense, editExpense, setExpenses, startSetExpenses };
