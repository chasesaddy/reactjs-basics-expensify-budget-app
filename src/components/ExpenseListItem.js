import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeraljs';

const ExpenseListItem = ( { id, description, amount, createdAt } ) => (
  <div>
    <h3>
      <Link to={ `${ id }/edit` }>{ description }</Link>
    </h3>
    <p>
      { numeral( amount / 100 ).format( '$0,0.00' ) } 
       -  
      { moment( createdAt ).format( 'MMMM Do, YYYY' ) }
    </p>
  </div>
);

export default ExpenseListItem;
