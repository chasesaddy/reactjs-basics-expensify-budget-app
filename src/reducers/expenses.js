// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = ( state = expensesReducerDefaultState, action ) => {
  switch ( action.type ) {
  case 'ADD_EXPENSE':
    return [
      ...state, action.expense
    ];
  case 'REMOVE_EXPENSE': 
    console.log( 'here' );
    console.log( action.id );
    return state.filter( ( oneExpense ) => oneExpense.id !== action.id );
    // Could have done destructuring. Didn't think of that.
    // return state.filter( ( { id } ) => id !== action.id );
  case 'EDIT_EXPENSE': 
    return state.map( ( oneExpense ) => {
      if ( oneExpense.id === action.id ) {
        return {
          ...oneExpense,
          ...action.updates
        };
      } else {
        return oneExpense;
      }
    });
  default:
    // will print filterReducer action.type
    console.log( action.type );
    return state;
  }
};

export default expensesReducer;
