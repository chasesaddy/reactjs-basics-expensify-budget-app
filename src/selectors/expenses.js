import moment from 'moment';

const getVisibleExpenses = ( expenses, { text, sortBy, startDate, endDate } ) => {
  return expenses.filter( ( solo ) => {
    
    const createdAtMoment = moment( solo.createdAt )

    const startDateMatched = startDate ? moment( startDate ).isSameOrBefore( createdAtMoment, 'day' ) : true;
    const endDateMatched = startDate ? moment( endDate ).isSameOrAfter( createdAtMoment, 'day' ) : true;
    const textMatched = textMatch( solo.description, text );

    return startDateMatched && endDateMatched && textMatched;
  }).sort( ( a, b ) => {
    if ( sortBy == 'date' ) {
      return b.createdAt > a.createdAt ? 1 : -1;
    } else if ( sortBy == 'amount' ) {
      return b.amount > a.amount ? 1 : -1;
    }
  });
};

const startDateMatch = ( objCreatedDate, startDate ) => {
  return ( typeof startDate !== 'number' || objCreatedDate >= startDate );
};

const endDateMatch = ( objCreatedDate, endDate ) => {
  return ( typeof endDate !== 'number' || objCreatedDate <= endDate );
};

const textMatch = ( objText, text ) => {
  return ( typeof text !== 'string' || objText.toLowerCase().includes( text.toLowerCase() ) );
};

export default getVisibleExpenses;
