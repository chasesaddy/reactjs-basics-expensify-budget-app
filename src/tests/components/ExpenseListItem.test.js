import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';

import expenses from '../fixtures/expenses';

test( 'should render one ExpenseListItem from fixture data', () => { 
  const example = expenses[ 0 ];
  const wrapper = shallow( <ExpenseListItem key={ example.id } { ...example } /> );

  expect( wrapper ).toMatchSnapshot();
} );
