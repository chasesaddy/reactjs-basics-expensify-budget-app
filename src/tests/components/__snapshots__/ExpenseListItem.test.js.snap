// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render one ExpenseListItem from fixture data 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="1/edit"
    >
      Gumming
    </Link>
  </h3>
  <p>
    $3.50
    -
    December 27th, 1969
  </p>
</div>
`;
